stl2pdf.1
.\" -*- nroff -*-
.\" stl2ps.1
.\" By: R.F. Smith <rsmith@xs4all.nl>
.\" $Date$
.Dd
.Os 
.Dt STL2PDF 1 CON
.Sh NAME
.Nm stl2pdf
.Nd Projects an STL file on the (x,y) plane after optionally rotating it and
renders a PDF figure of it. 
.Sh SYNOPSIS
.Nm 
infile
.Op Ar outfile
.Op x Ar degrees | y Ar degrees | z Ar degrees ...
.Sh DESCRIPTION
This manual page documents 
.Ql $Revision$
of the 
.Nm
program. This program's purpose is to render a globally illuminated grayscale
projection of an STL file as PDF output.

First, the STL file is parsed. Both binary and text STL file can be read, but
any color information in binary STL files is discarded since this is not
standardized. While STL files are supposed to contain closed solids, by design
no effort is made to try and repair gaps in the STL file. There simply isn't
enough geometric context to do this in a foolproof way.

Next, all the triangles in the file are rotated around the axis of the global
coordinate system, as specified by the sequence of rotation options given on
the command line.

Finally, the visible triangles (visibility is determined by a non-negative
z-component of the unit normal vector of the triangle) are sorted by depth,
projected on the (x,y) plane. So in the PDF picture you are looking along the
z-axis to the origin. The picture is scaled to fit in a maximum size of
100x100 mm and the triangles are drawn from back to front. The grayscale tint
of the triangle is also determined by the orientation of the triangle; running
from almost black for triangles turned 90 degrees to the line of sight to near
white for triangles parallel to the x,y plane.

If no output filename is given, the path and the
.Ql .stl 
extension are stripped from the input filename. The extension is replaced by the
.Ql .pdf
extension. So if no output filename is specified, the output file is written
in the current working directory. The PDF file is written to the output
filename.
.Sh EXIT STATUS
.Ex -std
.Sh DIAGNOSTICS
The following non-fatal warnings can be issued when the program cannot make
sense of command line arguments:
.Bl -tag -width "foo"
.It Unknown argument '...' ignored.
One of the command line transform arguments did not start with one of the
letters 'x', 'y', 'z', 'X', 'Y' or 'Z' and will be ignored. 
.It Argument '...' is not a number, ignored.
A non-numeric value was encountered among the command line arguments while a
number of degrees is expected as part of a rotation command. It will be ignored.
.El
.Pp
There are some errors which the program cannot ignore, and which will
terminate the program.
.Bl -tag -width "foo"
.It The file '...' cannot be read or parsed. Exiting.
If a given input file cannot be read, or if the file is not recognized as an
STL file, this error is produced, and the program is terminated with exit code
1.
.It Cannot write output file '...'.
An output file name was given on the command line, but the file in question
could not be opened for writing or written to. The program is terminated with
exit code 2.
.El
.Sh COMPATIBILITY
The 
.Nm
program requires the Python interpreter. It was written for version
2.7. Additionally, it requires the open source 
.Xr "Cairo" 
vector graphics library, as well as its Pyton bindings.
.Sh SEE ALSO
.Xr python 1 ,
.Xr stl2pov 1 ,
.Xr stl2ps 1
.Rs
.%B Python Programming Language - Official Website
.%U http://www.python.org/
.Re
.Rs
.%B Cairo
.%U http://cairographics.org/
.Re
.Sh HISTORY
The origin of this software was found in the desire of the author to render 3D
CAD models using the POV-ray raytracer. The first effort produced the
.Xr stl2pov 1
program, written in C. 
.Pp
Later the author converted that program to Python as a learning
experience. This resulted in a more re-usable version of the software to parse
STL files as a Python module. Having done that, adding a front-end to produce
PostScript output was relatively easy.
.Pp
With the 
.Xr "Cairo" 
library it was relatively easy to modify
.Xr stl2ps
to produce PDF output directly.
.Sh AUTHOR
This manual and the 
.Nm
software were written by 
.An Roland Smith Aq rsmith@xs4all.nl .
.Pp
The latest version of this program is available at:
.Lk http://rsmith.home.xs4all.nl/software/ 
.Sh LICENSE
To the extent possible under law, Roland Smith has waived all copyright and
related or neighboring rights to this manual. This work is published from the
Netherlands. See 
.Lk http://creativecommons.org/publicdomain/zero/1.0/
.Pp
The
.Nm
program itself is released under the two-clause BSD license given below;
.Pp
Copyright \(co 2012 R.F. Smith <rsmith@xs4all.nl>. All rights reserved.
.Pp
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
.Bl -tag -width "foo" -compact
.It 1.
Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.
.It 2.
Redistributions in binary form must reproduce the above copyright notice, this
list of conditions and the following disclaimer in the documentation and/or
other materials provided with the distribution.
.El
.Pp
THIS SOFTWARE IS PROVIDED BY AUTHOR AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS
OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" EOF


stl2pdf.py
#! /usr/bin/env python
# -*- python coding: utf-8 -*-
# Copyright © 2012 R.F. Smith <rsmith@xs4all.nl>. All rights reserved.
# $Date$
# 
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY AUTHOR AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.

'''Program for converting a view of an STL file into a PDF file.'''

import sys
import os
import stl
import xform
import cairo

name = ('stl2pdf [ver. ' + '$Revision$'[11:-2] + 
       '] ('+'$Date$'[7:-2]+')')

def usage():
    print name
    print "Usage: stl2pdf infile [outfile] [transform [transform ...]]"
    print "where [transform] is [x number|y number|z number]"

## This is the main program ##
# Process the command-line arguments
validargs = ['x', 'y', 'z', 'X', 'Y', 'Z']
if len(sys.argv) == 1:
    usage()
    sys.exit(0)
infile = sys.argv[1]
if len(sys.argv) < 3 or sys.argv[2] in validargs:
    outfile = None
    del sys.argv[:2]
else:
    outfile = sys.argv[2]
    del sys.argv[:3]
tr = xform.Xform()
while len(sys.argv) > 1:
    if not sys.argv[0] in validargs:
        print "Unknown argument '{}' ignored.".format(sys.argv[0])
        del sys.argv[0]
        continue
    try:
        ang = float(sys.argv[1])
        if sys.argv[0] in ['x','X']:
            tr.rotx(ang)
        elif sys.argv[0] in ['y','Y']:
            tr.roty(ang)
        else:
            tr.rotz(ang)
        del sys.argv[:2]
    except:
        print "Argument '{}' is not a number, ignored.".format(sys.argv[1])
        continue
# Open the file
try:
    stlobj = stl.Surface(infile)
except:
    print "The file '{}' cannot be read or parsed. Exiting.".format(sys.argv[1])
    sys.exit(1)
# Process the file
for result in stlobj.processfacets:
    print result
# Remove spaces from name
stlobj.name = stlobj.name.strip()
# Apply transformations
if tr.unity == False:
    stlobj.xform(tr)
# Calculate viewport and transformation
xmin, xmax, ymin, ymax, zmin, zmax = stlobj.extents()
pr = xform.Zpar(xmin, xmax, ymin, ymax)
# Prepare output.
if outfile == None:
    outbase = os.path.basename(infile)
    if outbase.endswith((".stl", ".STL")):
        outbase = outbase[:-4]
    outfile = outbase+".pdf"
#out = canvas.Canvas(outfile, (pr.w, pr.h), pageCompression=1)
out = cairo.PDFSurface(outfile, pr.w, pr.h)
ctx = cairo.Context(out)
ctx.set_line_cap(cairo.LINE_CAP_ROUND)
ctx.set_line_join(cairo.LINE_JOIN_ROUND)
ctx.set_line_width(0.25)

# Calculate the visible facets
vizfacets = [f for f in stlobj.facets if pr.visible(f.n)]
# Next, depth-sort the facets using the largest z-value of the three vertices.
vizfacets.sort(None, lambda f: max([f.v[0].z, f.v[1].z, f.v[2].z]))
# Project and illuminate the facets
pf = (stl.ProjectedFacet(f, pr) for f in vizfacets)
# Draw the triangles
for f in pf:
    path = ctx.new_path()
    ctx.move_to(f.x1, f.y1)
    ctx.line_to(f.x2, f.y2)
    ctx.line_to(f.x3, f.y3)
    ctx.close_path()
    ctx.set_source_rgb(f.gray, f.gray, f.gray)
    ctx.fill_preserve()
    ctx.stroke()
# Send output.
out.show_page()
out.finish()
